tokio-sync/src/task/atomic_task.rs

```diff
@@ -168,12 +168,23 @@ impl AtomicTask {
     /// tasks to be notified. One of the callers will win and have its task set,
     /// but there is no guarantee as to which caller will succeed.
     pub fn register_task(&self, task: Task) {
+        self.do_register(move || task)
+    }
+
+    fn do_register<F: FnOnce() -> Task>(&self, func: F) {
         debug!(" + register_task");
         match self.state.compare_and_swap(WAITING, REGISTERING, Acquire) {
             WAITING => {
                 unsafe {
                     // Locked acquired, update the waker cell
-                    self.task.with_mut(|t| *t = Some(task));
+                    self.task.with_mut(|t| {
+                        let should_update = (&*t).as_ref()
+                            .map(|prev| !prev.will_notify_current())
```

Comment:
I don't think this is correct. `register_task` could register a task that is *not* current.

There are a few options, but I'm not sure what is best.

Reply:
You mean that as the API is now, someone *could* call `atomic_task.register_task(some_other_task_because_reasons)`, and in the case, it should definitely register the task even if the current one would notify?

Reply:
Yep. I don’t remember why the api was added. It would be worth tracking down usage in Tokio.

Reply:
I've updated from a simple `FnOnce` generic to a more specialized trait, only check for `will_notify_current` in the `atomic.register()` case.